numpy/core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -995,34 +1006,52 @@ PyArray_GetStridedNumericCastFn(int aligned, npy_intp src_stride,
 #  if NPY_USE_UNALIGNED_ACCESS
                     if (src_stride == sizeof(@type1@) &&
                                 dst_stride == sizeof(@type2@)) {
-                        return &_aligned_contig_cast_@name1@_to_@name2@;
+                        return &_aligned_contig_cast_@name1@_to_@name2@@isa@;
                     }
                     else {
-                        return &_aligned_cast_@name1@_to_@name2@;
+                        return &_aligned_cast_@name1@_to_@name2@@isa@;
                     }
 #  else
                     if (src_stride == sizeof(@type1@) &&
                                 dst_stride == sizeof(@type2@)) {
                         return aligned ?
-                                    &_aligned_contig_cast_@name1@_to_@name2@ :
-                                    &_contig_cast_@name1@_to_@name2@;
+                                    &_aligned_contig_cast_@name1@_to_@name2@@isa@ :
+                                    &_contig_cast_@name1@_to_@name2@@isa@;
                     }
                     else {
-                        return aligned ? &_aligned_cast_@name1@_to_@name2@ :
-                                         &_cast_@name1@_to_@name2@;
+                        return aligned ? &_aligned_cast_@name1@_to_@name2@@isa@ :
+                                         &_cast_@name1@_to_@name2@@isa@;
```

Comment:
AVX/AVX2 unaligned accesses are fine as long as there is no annotation or anything that would cause the compiler to assume that pointers are aligned (AFAIK this is not the case in this part of the code). x86-64 processors historically supported this correctly but inefficiently (there was some penalty cycles due to crossing cache lines for example). All recent processors (at least 5 years old) implement this both correctly and efficiently (almost no overhead on contiguous data).

Unaligned loads (more precisely if `addr % sizeof(dest_type) != 0`) are technically undefined behavior in C. Thus, the code must use things like `memcpy` or `memmove` to perform a proper type punning. This is what the casting function appear to do. Using AVX/AVX2 does not change this.

In fact, it turns out GCC is able to optimize the `memmove` calls and replace them with unaligned AVX2 loads. Moreover, the above code works when it is executed on my machine. Thus, this unaligned case can benefit from AVX2 and is correctly optimized by GCC although most users should not deal with views.

```none
Unaligned: 1.29 ms ± 25.6 µs
Aligned: 1.3 ms ± 28.3
```

The aligned and unaligned version runs equally fast on my machine (with a non-statistically-significant difference).

I was wondering if the code would be fast and correct for non-contiguous unaligned views but this case is apparently currently forbidden by Numpy yet (I got an error message saying this when `np.view` is called).

Reply:
no, it isn't forbidden, maybe you got a wrong view incarnation, try:
```
arr_unaligned_noncontig = np.empty(10*1024**2 + 1, dtype="uint8")[1:].view(from_dt)[::2]
```